### PR TITLE
String formatted query updates

### PR DESCRIPTION
**Pull request by hex0punk**
This update adds the following patterns:

```yaml
    - pattern: |
        $QUERY = fmt.$F("...", ...)
        ...
        $OBJ.QueryContext($CTX, $QUERY, ...)
    - pattern: |
        $QUERY = fmt.$F("...", ...)
        ...
        $OBJ.QueryRow($QUERY)
    - pattern: |
        $QUERY = fmt.$F("...", ...)
        ...
        $OBJ.Exec($QUERY)
    - pattern: |
        $QUERY = fmt.$F("...", ...)
        ...
        $OBJ.QueryRow($CTX, $QUERY)
    - pattern: |
        $QUERY = fmt.$F("...", ...)
        ...
        $OBJ.QueryRowContext($CTX, $QUERY, ...)
```

These patterns allow the rule to now find cases such as the following which were previously missed/not detected by the rule:

```go
func postgresQueryFmt(r *http.Request) {
    ctx := context.Background()
    customerId := r.URL.Query().Get("id")
	// ruleid: string-formatted-query
	query := fmt.Sprintf("SELECT number, expireDate, cvv FROM creditcards WHERE customerId = %s", customerId)

    row, _ := postgresDb.QueryRow(ctx, query)
}

func dbQueryContextFmt(r *http.Request) {
	ctx := context.Background()
	customerId := r.URL.Query().Get("id")
	// ruleid: string-formatted-query
	query := fmt.Sprintf("SELECT number, expireDate, cvv FROM creditcards WHERE customerId = %s", customerId)
	row, _ := db.QueryContext(ctx, query)
}

func dbQueryRowContextFmt(r *http.Request) {
	ctx := context.Background()
	customerId := r.URL.Query().Get("id")
	// ruleid: string-formatted-query
	query := fmt.Sprintf("SELECT number, expireDate, cvv FROM creditcards WHERE customerId = %s", customerId)

	row, _ := db.QueryRowContext(ctx, query)
}

func dbQuery4(r *http.Request, username string) {
    // ruleid: string-formatted-query
    query := fmt.Sprintf("%s AND INSERT into users (username, password)", username)
    _, err = db.Exec(query)
    if err != nil {
        http.Error("mistake")
    }
}

func dbQuery5(r *http.Request, username string, password string) {
	// ruleid: string-formatted-query
	query := fmt.Sprintf("INSERT into users (username, password) VALUES(%s, %s)", username, password)
	_, err = db.QueryRow(query)
	if err != nil {
		http.Error("mistake")
	}
}
```

After adding the new rules, the semgrep tests reported duplicates. I updated the rule to remove rules that were causing duplicates given that the new rule finds those cases. I believe this also helps make the rule more accurate, as the use of `$FXN($QUERY)` as a metapattern could easily interpret $FXN  as an issue where it could very easily be sanitizer. Instead, using `fmt.$F` targets functions in the `fmt` package more commonly used to insert input without sanitization.

### CODE REVIEW

**Comment by LewisArdern** [comment]
This rule is in really bad shape and adding more patterns at this time without a big refactor would lead to a lot of more false positives.

I think we should at least try to remove 'strings' from the original sources would make my happiness a little higher

**Comment by hex0punk** [comment]
Agreed on the state of the rule, but unsure what you mean by removing 'strings' ? In any case, I can help with the refractor, maybe something along these lines to begin with (this is not complete yet). Let me know your thoughts 

```yaml
  - pattern-either:
    - pattern: $OBJ.Exec($X + ...)
    - pattern: $OBJ.ExecContext($CTX, $X + ...)
    - pattern: $OBJ.Query($X + ...)
    - pattern: $OBJ.QueryContext($CTX, $X + ...)
    - pattern: $OBJ.QueryRow($X + ...)
    - pattern: $OBJ.QueryRow($CTX, $X + ...)
    - pattern: $OBJ.QueryRowContext($CTX, $X + ...)
    - pattern: $OBJ.Exec(fmt.$P("...", ...))
    - pattern: $OBJ.ExecContext($CTX, fmt.$P("...", ...))
    - pattern: $OBJ.Query(fmt.$P("...", ...))
    - pattern: $OBJ.QueryContext($CTX, fmt.$P("...", ...))
    - pattern: $OBJ.QueryRow(fmt.$P("...", ...))
    - pattern: $OBJ.QueryRow($CTX, fmt.$U("...", ...))
    - pattern: $OBJ.QueryRowContext($CTX, fmt.$P("...", ...))
    - patterns:
        - pattern-either:
            - pattern: $QUERY = fmt.Fprintf($F, "$SQLSTR", ...)
            - pattern: $QUERY = fmt.Sprintf("$SQLSTR", ...)
            - pattern: $QUERY = fmt.Printf("$SQLSTR", ...)
            - pattern: $QUERY = $X + ...
        - pattern-either:
            - pattern-inside: |
                ...
                $OBJ.Query($QUERY, ...)
            - pattern-inside: |
                ...
                $OBJ.ExecContext($CTX, $QUERY, ...)
            - pattern-inside: |
                ...
                $OBJ.Exec($QUERY, ...)
            - pattern-inside: |
                ...
                $OBJ.QueryRow($CTX, $QUERY)
            - pattern-inside: |
                ...
                $OBJ.QueryRowContext($CTX, $QUERY, ...)
```

**Comment by hex0punk** [comment]
I did some refactoring of the rule and pushed the change. Let me know what you think could be improved @LewisArdern

**Comment by LewisArdern** [comment]
@kurt-r2c any chance u can take a look at this?

File: go/lang/security/audit/database/string-formatted-query.yaml [hunk 1]
```diff
@@ -53,117 +53,34 @@ rules:
...
-        $OBJ.QueryRowContext($CTX, $QUERY, ...)
+    - patterns:
+        - pattern-either:
+            - pattern: $QUERY = fmt.Fprintf($F, "$SQLSTR", ...)
+            - pattern: $QUERY = fmt.Sprintf("$SQLSTR", ...)
+            - pattern: $QUERY = fmt.Printf("$SQLSTR", ...)
+            - pattern: $QUERY = $X + ...
+        - pattern-either:
+            - pattern-inside: |
+                ...
```
Comment (kurt-r2c) [inline go/lang/security/audit/database/string-formatted-query.yaml]: leading ellipsis statements are *really* slow - this will hurt the performance of this rule a lot. Some kind of upper bound on this ellipsis will improve the performance of this rule significantly - perhaps a function definition? Go is opinionated enough about program structure that this should be an acceptable constraint.

**Review by kurt-r2c — CHANGES_REQUESTED** [review]

File: go/lang/security/audit/database/string-formatted-query.yaml (hunk 1 above)
Comment (hex0punk) [reply go/lang/security/audit/database/string-formatted-query.yaml]: Good to know. I tested the lastest commit I pushed and I am getting the expected results. Let me know what you think

**Review by kurt-r2c — DISMISSED** [review]
much better

File: go/lang/security/audit/database/string-formatted-query.yaml [hunk 2]
```diff
@@ -53,117 +53,48 @@ rules:
...
+        - pattern-either:
+            - pattern: $QUERY = fmt.Fprintf($F, "$SQLSTR", ...)
+            - pattern: $QUERY = fmt.Sprintf("$SQLSTR", ...)
+            - pattern: $QUERY = fmt.Printf("$SQLSTR", ...)
+            - pattern: $QUERY = $X + ...
+        - pattern-either:
+            - pattern-inside: |
+                func $FUNC(...) {
+                    ...
+                    $OBJ.Query($QUERY, ...)
```
Comment (kurt-r2c) [inline go/lang/security/audit/database/string-formatted-query.yaml:68]: This will perform much better - now that we're scoping to a function, we probably need a trailing ellipsis to make sure we don't miss e.g.

```go
func postgresQueryFmt(r *http.Request) {
    ctx := context.Background()
    customerId := r.URL.Query().Get("id")
	// ruleid: string-formatted-query
	query := fmt.Sprintf("SELECT number, expireDate, cvv FROM creditcards WHERE customerId = %s", customerId)

    row, _ := postgresDb.QueryRow(ctx, query)
    someotherResult := doSomethingElse()
}
```

**Review by kurt-r2c — CHANGES_REQUESTED** [review]
I think there's one more small fix we need to make and then this is OK to merge.

File: go/lang/security/audit/database/string-formatted-query.yaml (hunk 2 above)
Comment (hex0punk) [reply go/lang/security/audit/database/string-formatted-query.yaml:68]: Makes sense to me. Added.

**Comment by hex0punk** [comment]
Just making sure this is ready now, @kurt-r2c

**Review by kurt-r2c — APPROVED** [review]
